File: notebot/bot.py
import discord
from discord.ext import commands
import asyncio, os, time
from datetime import datetime
from dotenv import load_dotenv

from transcriber import transcribe_audio
from summarizer import summarize_transcript
from formatter import build_embed
from google_docs import create_meeting_doc
from airtable_tasks import log_tasks
from db import init_db, save_meeting

load_dotenv()
init_db()

intents = discord.Intents.default()
intents.message_content = True
GUILD_ID = int(os.getenv("DISCORD_GUILD_ID", 0))
bot = commands.Bot(command_prefix="!", intents=intents, debug_guilds=[GUILD_ID])

TRIGGER_USERS = {236708079872376834, 1074610938684121138}  # Mike, Alex

connections = {}          # guild_id → voice_client
start_times = {}          # guild_id → unix timestamp
channel_names = {}        # guild_id → voice channel name
recording_channels = {}   # guild_id → channel_id being auto-recorded

@bot.event
async def on_ready():
    await bot.sync_commands()
    print(f"✅ NoteBot online as {bot.user}")
    print(f"📊 Dashboard: http://localhost:{os.getenv('DASHBOARD_PORT', 8080)}")

@bot.event
async def on_voice_state_update(member, before, after):
    guild = member.guild
    guild_id = guild.id

    # Someone joined or moved into a channel
    if after.channel is not None and before.channel != after.channel:
        channel = after.channel
        member_ids = {m.id for m in channel.members}
        if TRIGGER_USERS.issubset(member_ids) and guild_id not in connections:
            channel_id = os.getenv("DISCORD_TEXT_CHANNEL_ID")
            text_channel = (
                guild.get_channel(int(channel_id)) if channel_id
                else guild.system_channel or next(
                    (c for c in guild.text_channels if c.permissions_for(guild.me).send_messages), None
                )
            )
            if text_channel is None:
                return
            vc = await channel.connect()
            connections[guild_id] = vc
            start_times[guild_id] = time.time()
            channel_names[guild_id] = channel.name
            recording_channels[guild_id] = channel.id
            vc.start_recording(
                discord.sinks.MP3Sink(),
                recording_finished,
                text_channel,
                guild_id
            )
            await text_channel.send(
                f"🎙️ Auto-recording **{channel.name}** — both members present."
            )

    # Someone left or moved out of the recording channel
    if (before.channel is not None
            and guild_id in connections
            and before.channel.id == recording_channels.get(guild_id)):
        remaining_ids = {m.id for m in before.channel.members}
        if not TRIGGER_USERS.intersection(remaining_ids):
            vc = connections.pop(guild_id)
            recording_channels.pop(guild_id, None)
            vc.stop_recording()
            await vc.disconnect()

@bot.slash_command(name="join", description="Join your voice channel and start recording")
async def join(ctx):
    if not ctx.author.voice:
        return await ctx.respond("❌ You need to be in a voice channel first.", ephemeral=True)

    await ctx.defer()
    channel = ctx.author.voice.channel
    vc = await channel.connect()
    connections[ctx.guild.id] = vc
    start_times[ctx.guild.id] = time.time()
    channel_names[ctx.guild.id] = channel.name

    vc.start_recording(
        discord.sinks.MP3Sink(),
        recording_finished,
        ctx.channel,
        ctx.guild.id
    )

    await ctx.followup.send(
        f"🎙️ Recording **{channel.name}** — type `/leave` when the meeting ends."
    )

async def recording_finished(sink, text_channel, guild_id):
    duration = int((time.time() - start_times.pop(guild_id, time.time())) / 60)
    voice_channel = channel_names.pop(guild_id, "Unknown")

    status_msg = await text_channel.send("⏳ Transcribing audio...")

    full_transcript_lines = []
    speaker_names = []

    for user_id, audio in sink.audio_data.items():
        try:
            user = await bot.fetch_user(user_id)
            speaker_names.append(user.display_name)
            audio_bytes = audio.file.read()
            text = await asyncio.to_thread(transcribe_audio, audio_bytes)
            if text.strip():
                full_transcript_lines.append(f"{user.display_name}: {text.strip()}")
        except Exception as e:
            print(f"Transcription error for user {user_id}: {e}")

    if not full_transcript_lines:
        await status_msg.edit(content="❌ No audio detected — nothing to summarize.")
        return

    transcript = "\n".join(full_transcript_lines)
    await status_msg.edit(content="🧠 Summarizing with Nemotron...")

    try:
        summary = await summarize_transcript(transcript)
    except Exception as e:
        await status_msg.edit(content=f"❌ Summarization failed: {e}")
        return

    meeting_date = datetime.utcnow().strftime("%Y-%m-%d")
    meeting_title = f"AMP Titans Meeting — {meeting_date}"

    # 1. Google Doc
    doc_url = ""
    try:
        await status_msg.edit(content="📄 Creating Google Doc...")
        doc_url = await asyncio.to_thread(
            create_meeting_doc, meeting_title, summary, transcript, speaker_names
        )
    except Exception as e:
        print(f"Google Docs error: {e}")

    # 2. Airtable tasks
    try:
        await status_msg.edit(content="📌 Logging tasks to Airtable...")
        await asyncio.to_thread(
            log_tasks, summary.get("action_items", []), meeting_date, meeting_title
        )
    except Exception as e:
        print(f"Airtable error: {e}")

    # 3. Save to SQLite
    meeting_id = save_meeting({
        "date": meeting_date,
        "channel": voice_channel,
        "duration_minutes": duration,
        "speakers": speaker_names,
        "summary": summary.get("summary", ""),
        "topics": summary.get("topics", []),
        "decisions": summary.get("decisions", []),
        "action_items": summary.get("action_items", []),
        "blockers": summary.get("blockers", []),
        "transcript": transcript,
        "google_doc_url": doc_url
    })

    # 4. Discord embed + doc link
    embed = build_embed(summary, speaker_names, duration, doc_url, meeting_id)
    await status_msg.delete()
    await text_channel.send(embed=embed)
    if doc_url:
        await text_channel.send(f"📄 **Google Doc:** {doc_url}")

@bot.slash_command(name="leave", description="Stop recording and generate meeting notes")
async def leave(ctx):
    if ctx.guild.id not in connections:
        return await ctx.respond("❌ I'm not recording right now.", ephemeral=True)

    await ctx.defer()
    vc = connections.pop(ctx.guild.id)
    recording_channels.pop(ctx.guild.id, None)
    vc.stop_recording()
    await vc.disconnect()
    await ctx.followup.send("✅ Stopped. Processing notes now...")

bot.run(os.getenv("DISCORD_BOT_TOKEN"))

File: notebot/bot.js
'use strict';

require('dotenv').config();
const {
  Client, GatewayIntentBits, EmbedBuilder,
  REST, Routes, SlashCommandBuilder
} = require('discord.js');
const { joinVoiceChannel, EndBehaviorType } = require('@discordjs/voice');
const prism = require('prism-media');
const { createWriteStream, createReadStream, unlinkSync, existsSync, statSync } = require('fs');
const { pipeline } = require('stream');
const axios = require('axios');
const FormData = require('form-data');

const TRIGGER_USERS = new Set([
  '236708079872376834',   // Mike
  '1074610938684121138'   // Alex
]);

const GUILD_ID        = process.env.DISCORD_GUILD_ID;
const TEXT_CHANNEL_ID = process.env.DISCORD_TEXT_CHANNEL_ID;
const API_BASE        = `http://localhost:${process.env.DASHBOARD_PORT || 8080}`;

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildVoiceStates,
    GatewayIntentBits.MessageContent,
  ]
});

// — State —
let voiceConn         = null;
let startTime         = null;
let recordingChanId   = null;
let recordingChanName = '';
let recordingText     = null;
let speakerFiles      = new Map(); // userId → { filename, displayName, rawStream, writeStream }

// ─── Audio helpers ────────────────────────────────────────────────────────────

function subscribeUser(receiver, userId, displayName) {
  if (speakerFiles.has(userId)) return;

  const filename    = `rec_${userId}_${Date.now()}.pcm`;
  const rawStream   = receiver.subscribe(userId, { end: EndBehaviorType.Manual });
  const decoder     = new prism.opus.Decoder({ frameSize: 960, channels: 2, rate: 48000 });
  const writeStream = createWriteStream(filename);

  pipeline(rawStream, decoder, writeStream, err => {
    if (err && err.code !== 'ERR_STREAM_DESTROYED') {
      console.error(`[audio] error for ${displayName}:`, err.message);
    }
  });

  speakerFiles.set(userId, { filename, displayName, rawStream, writeStream });
  console.log(`[rec] subscribed to ${displayName}`);
}

// ─── Recording lifecycle ──────────────────────────────────────────────────────

async function startRecording(voiceChannel, txtChannel, sendMessage = true) {
  if (voiceConn) return;

  recordingText     = txtChannel;
  startTime         = Date.now();
  recordingChanName = voiceChannel.name;
  recordingChanId   = voiceChannel.id;
  speakerFiles      = new Map();

  voiceConn = joinVoiceChannel({
    channelId:       voiceChannel.id,
    guildId:         voiceChannel.guild.id,
    adapterCreator:  voiceChannel.guild.voiceAdapterCreator,
    selfDeaf:        false,
    selfMute:        true,
  });

  const receiver = voiceConn.receiver;

  // Subscribe to everyone already in the channel
  for (const [uid, member] of voiceChannel.members) {
    if (uid === client.user.id) continue;
    subscribeUser(receiver, uid, member.displayName);
  }

  // Subscribe as new people start speaking
  receiver.speaking.on('start', uid => {
    const member = voiceChannel.members.get(uid);
    if (member && uid !== client.user.id) {
      subscribeUser(receiver, uid, member.displayName);
    }
  });

  if (sendMessage) {
    await txtChannel.send(`🎙️ Recording **${voiceChannel.name}** — type \`/leave\` when done.`);
  }
}

async function stopRecording() {
  if (!voiceConn) return;

  const duration  = Math.round((Date.now() - startTime) / 60000);
  const txtChan   = recordingText;
  const chanName  = recordingChanName;

  // Signal end of all raw streams (push null, don't destroy yet)
  for (const [, { rawStream }] of speakerFiles) {
    try { rawStream.push(null); } catch (_) {}
  }

  // Wait for all write streams to finish flushing (up to 8 seconds)
  await Promise.all([...speakerFiles.values()].map(({ writeStream }) =>
    new Promise(resolve => {
      if (writeStream.writableEnded || writeStream.closed) return resolve();
      writeStream.once('finish', resolve);
      writeStream.once('close', resolve);
      setTimeout(resolve, 8000);
    })
  ));

  voiceConn.destroy();
  voiceConn       = null;
  recordingChanId = null;

  await processRecording(duration, chanName, txtChan);
}

// ─── Processing ───────────────────────────────────────────────────────────────

async function processRecording(duration, chanName, txtChan) {
  const statusMsg = await txtChan.send('⏳ Transcribing audio...');

  try {
    const form = new FormData();
    form.append('channel',  chanName);
    form.append('duration', String(duration));

    let hasAudio = false;
    for (const [userId, { filename, displayName }] of speakerFiles) {
      if (existsSync(filename) && statSync(filename).size > 0) {
        console.log(`[process] including ${displayName} — ${statSync(filename).size} bytes`);
        form.append('audio_files',   createReadStream(filename), {
          filename:    `${userId}.pcm`,
          contentType: 'audio/pcm',
        });
        form.append('speaker_names', displayName);
        hasAudio = true;
      } else {
        console.log(`[process] skipping ${displayName} — file missing or empty`);
      }
    }

    if (!hasAudio) {
      await statusMsg.edit('❌ No audio recorded.');
      return;
    }

    const resp = await axios.post(`${API_BASE}/process`, form, {
      headers: form.getHeaders(),
      timeout: 600_000,       // 10 min — allow for long transcription
      maxContentLength: Infinity,
      maxBodyLength:    Infinity,
    });

    const result = resp.data;

    // Clean up PCM files
    for (const [, { filename }] of speakerFiles) {
      if (existsSync(filename)) unlinkSync(filename);
    }

    const embed = buildEmbed(result);
    await statusMsg.delete();
    await txtChan.send({ embeds: [embed] });
    if (result.google_doc_url) {
      await txtChan.send(`📄 **Google Doc:** ${result.google_doc_url}`);
    }

  } catch (err) {
    console.error('[process] error:', err.message);
    await statusMsg.edit(`❌ Processing failed: ${err.message}`);

    // Clean up on failure too
    for (const [, { filename }] of speakerFiles) {
      if (existsSync(filename)) try { unlinkSync(filename); } catch (_) {}
    }
  }
}

function buildEmbed(r) {
  const embed = new EmbedBuilder()
    .setTitle('📋 Meeting Notes')
    .setDescription(r.summary || 'No summary available.')
    .setColor(0x5865F2)
    .setTimestamp();

  if (r.topics?.length)
    embed.addFields({ name: '🗂️ Topics',       value: r.topics.map(t => `• ${t}`).join('\n'),       inline: false });
  if (r.decisions?.length)
    embed.addFields({ name: '✅ Decisions',     value: r.decisions.map(d => `• ${d}`).join('\n'),    inline: false });
  if (r.action_items?.length)
    embed.addFields({ name: '📌 Action Items',  value: r.action_items.map(a => `• ${a}`).join('\n'), inline: false });
  if (r.blockers?.length)
    embed.addFields({ name: '🚧 Blockers',      value: r.blockers.map(b => `• ${b}`).join('\n'),     inline: false });

  const links = [];
  if (r.google_doc_url) links.push(`[📄 Google Doc](${r.google_doc_url})`);
  links.push(`[📊 Dashboard](http://localhost:8080/meeting/${r.meeting_id})`);
  embed.addFields({ name: '🔗 Links', value: links.join('  |  '), inline: false });
  embed.setFooter({ text: `AMP Titans NoteBot • ${(r.speakers || []).join(', ')} • ${r.duration} min` });

  return embed;
}

// ─── Auto-trigger ─────────────────────────────────────────────────────────────

client.on('voiceStateUpdate', async (oldState, newState) => {
  const memberId = newState.id || oldState.id;
  if (memberId === client.user?.id) return;

  const guild = newState.guild;

  // Someone joined a channel — check if both trigger users are now present
  if (newState.channelId && newState.channelId !== oldState.channelId) {
    const channel   = newState.channel;
    const memberIds = new Set([...channel.members.keys()]);

    const allPresent = [...TRIGGER_USERS].every(id => memberIds.has(id));
    if (allPresent && !voiceConn) {
      const txtChan = TEXT_CHANNEL_ID
        ? guild.channels.cache.get(TEXT_CHANNEL_ID)
        : guild.systemChannel;
      if (txtChan) await startRecording(channel, txtChan, true);
    }
  }

  // Someone left the recording channel — stop when both trigger users are gone
  if (oldState.channelId && oldState.channelId === recordingChanId) {
    const oldChan    = oldState.channel;
    const remaining  = new Set([...(oldChan?.members?.keys() || [])]);
    const anyPresent = [...TRIGGER_USERS].some(id => remaining.has(id));
    if (!anyPresent && voiceConn) await stopRecording();
  }
});

// ─── Slash commands ───────────────────────────────────────────────────────────

client.on('interactionCreate', async interaction => {
  if (!interaction.isChatInputCommand()) return;

  if (interaction.commandName === 'join') {
    if (!interaction.member.voice.channel) {
      return interaction.reply({ content: '❌ You need to be in a voice channel first.', ephemeral: true });
    }
    await interaction.deferReply();
    await startRecording(interaction.member.voice.channel, interaction.channel, false);
    await interaction.editReply(`🎙️ Recording **${recordingChanName}** — type \`/leave\` when done.`);
  }

  if (interaction.commandName === 'leave') {
    if (!voiceConn) {
      return interaction.reply({ content: '❌ Not recording right now.', ephemeral: true });
    }
    await interaction.deferReply();
    await interaction.editReply('✅ Stopped. Processing notes...');
    await stopRecording();
  }
});

// ─── Startup ──────────────────────────────────────────────────────────────────

client.once('ready', async () => {
  console.log(`✅ NoteBot online as ${client.user.tag}`);
  console.log(`📊 Dashboard: ${API_BASE}`);

  const commands = [
    new SlashCommandBuilder().setName('join').setDescription('Join your voice channel and start recording'),
    new SlashCommandBuilder().setName('leave').setDescription('Stop recording and generate meeting notes'),
  ].map(c => c.toJSON());

  const rest = new REST({ version: '10' }).setToken(process.env.DISCORD_BOT_TOKEN);
  try {
    await rest.put(Routes.applicationGuildCommands(client.user.id, GUILD_ID), { body: commands });
    console.log('✅ Slash commands registered');
  } catch (err) {
    console.error('Failed to register commands:', err.message);
  }
});

client.login(process.env.DISCORD_BOT_TOKEN).catch(err => {
  console.error('Login failed:', err.message);
  process.exit(1);
});
